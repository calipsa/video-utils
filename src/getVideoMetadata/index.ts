import { promises as fsPromises } from 'fs'

import * as tmp from 'tmp-promise'

import getMetadata from './getMetadata'

type Input = Buffer | string

export default async (input: Input) => {
  if (typeof input === 'string') {
    return getMetadata(input)
  }

  const { path, cleanup } = await tmp.file()
  try {
    await fsPromises.writeFile(path, input, 'binary')
    return await getMetadata(path)
  } finally {
    cleanup()
  }
}
