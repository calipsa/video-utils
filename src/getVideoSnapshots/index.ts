import * as tmp from 'tmp-promise'
import type { FfmpegCommand } from 'fluent-ffmpeg'

import type { Input } from '../types'
import writePolymorphic from '../utils/writePolymorphic'
// import noop from '../utils/noop'

import getImageBuffers from './getImageBuffers'

interface Options {
  input: Input,
  rate: number,
  format: 'jpg' | 'png',
  logError: (...args: any[]) => void,
  configureFfmpeg?: (ffmpegCommand: FfmpegCommand) => void,
}

export default async ({
  input,
  rate = 1,
  format = 'jpg',
  // logError = noop,
  configureFfmpeg,
}: Options) => {
  if (typeof input === 'string') {
    // eslint-disable-next-line no-return-await
    return await getImageBuffers(input, rate, format, configureFfmpeg)
  }

  const { path, cleanup } = await tmp.file()
  try {
    await writePolymorphic(path, input)
    return await getImageBuffers(path, rate, format, configureFfmpeg)
  } finally {
    await cleanup()
  }
}
