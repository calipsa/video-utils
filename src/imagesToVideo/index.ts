import { Writable } from 'stream'

import { ffmpeg } from 'ffmpeg-stream'

import type { PathOrReadable } from '../types'
import streamImages from './streamImages'

interface Params {
  images: PathOrReadable[],
  fps: number,
}

export default ({
  images,
  fps,
}: Params) => {
  const converter = ffmpeg()

  const input = converter.input({
    f: 'image2pipe',
    r: fps,
  }) as Writable

  // const output = converter.output({
  //   f: 'mpegts',
  //   vcodec: 'mpeg2video',
  //   tune: 'zerolatency',
  //   preset: 'ultrafast',
  //   'q:v': 1,
  // })

  const output = converter.output({
    f: 'avi',
    vcodec: 'wmv2',
    // buffer: true,
  })

  streamImages(images, input)
  converter.run()

  return output
}
