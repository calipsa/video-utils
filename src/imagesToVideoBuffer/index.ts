import type { Input } from '../types'

import convert from './convert'

interface Params {
  images: Input[],
  fps: number,
}

export default ({
  images,
  fps,
}: Params) => convert(images, fps)
