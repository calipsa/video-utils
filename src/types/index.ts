import { Readable } from 'stream'

export type PathOrReadable = string | Readable
export type Input = PathOrReadable | Buffer
