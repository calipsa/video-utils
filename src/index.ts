import getVideoSnapshots from './getVideoSnapshots'
import imagesToVideo from './imagesToVideo'

export {
  getVideoSnapshots,
  imagesToVideo,
}
