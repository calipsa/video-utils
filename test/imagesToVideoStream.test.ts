import fs from 'fs'
import Path from 'path'
import { streamToBuffer } from '@calipsa/stream-utils'
import { identity } from 'lodash'

import { imagesToVideoStream } from '../src'
import getHash from './utils/getHash'

const formats = [
  'wmv',
  'mp4',
] as const

const IMAGES_DIR = Path.join(__dirname, 'images')
const imageFileNames = fs.readdirSync(IMAGES_DIR).filter(name => name.endsWith('.jpg'))
const imagePaths = imageFileNames.sort().map(n => Path.join(IMAGES_DIR, n))

const pathToInput = {
  string: identity,
  stream: (path: string) => fs.createReadStream(path),
  buffer: (path: string) => fs.promises.readFile(path, null)
} as const

describe('Images to video stream', () => {
  beforeEach(() => {
    jest.setTimeout(120000)
  })

  it('Should convert images to video', async () => {
    const fileNames = await fs.promises.readdir(IMAGES_DIR)
    for (const format of formats) {
      const types = Object.keys(pathToInput)
      const bufferPromises = types.map(async (type) => {
        const f = pathToInput[type]
        const imagePromises = imagePaths.map(f) as any[]
        const images = await Promise.all(imagePromises)
        const stream = imagesToVideoStream({
          images,
          format,
          fps: 10,
        })
        return await streamToBuffer(stream)
      })

      const firstBuffer = await bufferPromises[0]
      // await fs.promises.writeFile(`./video.${format}`, firstBuffer)

      const promises = bufferPromises.map(p => p.then(getHash))

      const first = await promises[0]
      expect(first).toMatchSnapshot(format)
  
      for (let i = 1; i < types.length; ++i) {
        const v = await promises[i]
        expect(v).toEqual(first)
      }  
    }
  })
})
