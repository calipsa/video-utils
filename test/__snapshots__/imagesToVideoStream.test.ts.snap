// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Images to video stream Should convert images to video: mp4 1`] = `"jE/kbhnyiz80ABTqh0+5ttxUyVxi0tCHE9pXwOjEgOg="`;

exports[`Images to video stream Should convert images to video: wmv 1`] = `"gzR9E/1x/BPCMIXDu+JNMxTLJB3d+buVDN9Kw+roFYg="`;
